middleware/commandhandler/scheduler/middleware.go

```diff
@@ -82,6 +52,302 @@ func (c *command) ExecuteAt() time.Time {
+	if cmd, ok := cmd.(eh.CommandIDer); ok {
+		id = cmd.CommandID()
+	} else {
+		id = uuid.New()
+	}
+
+	pc := &PersistedCommand{
+		ID:         id,
+		IDStr:      id.String(),
+		RawCommand: b,
+		ExecuteAt:  executeAt,
+	}
+
+	if err := s.repo.Save(context.Background(), pc); err != nil {
+		return uuid.Nil, &Error{
+			Err:     fmt.Errorf("could not persist command: %w", err),
+			Ctx:     ctx,
+			Command: cmd,
+		}
+	}
+
+	select {
+	case s.cmdCh <- &scheduledCommand{id, ctx, cmd, executeAt}:
+	default:
+		return uuid.Nil, &Error{
+			Err:     fmt.Errorf("command queue full"),
+			Ctx:     ctx,
+			Command: cmd,
+		}
+	}
+
+	return pc.ID, nil
+}
+
+// Commands returns all scheduled commands.
+func (s *Scheduler) Commands(ctx context.Context) ([]*PersistedCommand, error) {
+	entities, err := s.repo.FindAll(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not load scheduled commands: %w", err)
+	}
+
+	commands := make([]*PersistedCommand, len(entities))
+
+	for i, entity := range entities {
+		c, ok := entity.(*PersistedCommand)
+		if !ok {
+			return nil, fmt.Errorf("command is not schedulable: %T", entity)
+		}
+
+		if c.Command, c.Context, err = s.codec.UnmarshalCommand(ctx, c.RawCommand); err != nil {
+			return nil, fmt.Errorf("could not unmarshal command: %w", err)
+		}
+
+		c.RawCommand = nil
+
+		if c.IDStr != "" {
+			id, err := uuid.Parse(c.IDStr)
+			if err != nil {
+				return nil, fmt.Errorf("could not parse command ID: %w", err)
+			}
+
+			c.ID = id
+		}
+
+		commands[i] = c
+	}
+
+	return commands, nil
+}
+
+// CancelCommand cancels a scheduled command.
+func (s *Scheduler) CancelCommand(ctx context.Context, id uuid.UUID) error {
+	s.cancelSchedulingMu.Lock()
+	defer s.cancelSchedulingMu.Unlock()
+
+	cancel, ok := s.cancelScheduling[id]
+	if !ok {
+		return fmt.Errorf("command %s not scheduled", id)
+	}
+
+	close(cancel)
+
+	return nil
+}
+
+func (s *Scheduler) loadCommands() error {
+	commands, err := s.Commands(context.Background())
+	if err != nil {
+		return fmt.Errorf("could not load scheduled commands: %w", err)
+	}
+
+	for _, pc := range commands {
+		sc := &scheduledCommand{
+			id:        pc.ID,
+			ctx:       pc.Context,
+			cmd:       pc.Command,
+			executeAt: pc.ExecuteAt,
+		}
+
+		select {
+		case s.cmdCh <- sc:
+		default:
+			return fmt.Errorf("could not schedule command: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Scheduler) run() {
+	var wg sync.WaitGroup
+
+loop:
+	for {
+		select {
+		case <-s.cctx.Done():
+			break loop
+		case sc := <-s.cmdCh:
+			wg.Add(1)
+
+			s.cancelSchedulingMu.Lock()
+			cancel := make(chan struct{})
+			s.cancelScheduling[sc.id] = cancel
+			s.cancelSchedulingMu.Unlock()
+
+			go func(cancel chan struct{}) {
+				defer wg.Done()
+
+				t := time.NewTimer(time.Until(sc.executeAt))
+				defer t.Stop()
+
+				select {
+				case <-s.cctx.Done():
+					// Stop without removing persisted cmd.
+				case <-t.C:
+					if err := s.h.HandleCommand(sc.ctx, sc.cmd); err != nil {
```

Comment:
Maybe "lock" the persisted command, to avoid multiple instances trying to issue the same command.

Reply:
Will be resolved later, added an issue for it (#364).